crypto/cipher_extra/cipher_test.cc

```diff
@@ -538,6 +540,161 @@ TEST(CipherTest, CAVP_TDES_ECB) {
+  ASSERT_TRUE(t->GetBytes(&tag, "tag"));
+
+  WycheproofResult result;
+  ASSERT_TRUE(GetWycheproofResult(t, &result));
+
+  // Allocate a ctx object
+  const EVP_CIPHER *cipher = EVP_chacha20_poly1305();
+  EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+  ASSERT_TRUE(ctx);
+
+  // Initialize without the key/iv
+  int res;
+  res = EVP_EncryptInit_ex(ctx, cipher, NULL, NULL, NULL);
+  ASSERT_EQ(res, 1);
+
+  // Set the IV size
+  int iv_size_val = std::stoi(iv_size) / 8;
+  res = EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, iv_size_val, NULL);
+  int is_invalid_iv_size = iv_size_val != 12;
+  if (is_invalid_iv_size && !result.IsValid()) {
+    // Check that we correctly reject the IV and skip.
+    // Otherwise, the cipher will simply just use IV=0.
+    ASSERT_EQ(res, 0);
+    t->SkipCurrent();
+    EVP_CIPHER_CTX_free(ctx);
+    return;
+  } else {
+    ASSERT_EQ(res, 1);
+  }
+
+  // Initialize with the key/iv
+  res = EVP_EncryptInit_ex(ctx, cipher, NULL, key.data(), iv.data());
+
+  // Insert AAD
+  int out_len = 0;
+  res = EVP_EncryptUpdate(ctx, NULL, &out_len, aad.data(), aad.size());
+  ASSERT_EQ(res, 1);
+  ASSERT_EQ(out_len, (int) aad.size());
+
+  // Insert plaintext
+  auto *computed_ct = (uint8_t *) OPENSSL_malloc(ct.size());
+  if (msg.size() > 0) {
+    out_len = 0;
+    res = EVP_EncryptUpdate(ctx, computed_ct, &out_len, msg.data(),
+                            msg.size());
+    ASSERT_EQ(res, 1);
+    ASSERT_EQ(out_len, (int) msg.size());
+  }
+
+  // Finish the cipher
+  out_len = 0;
+  res = EVP_EncryptFinal(ctx, NULL, &out_len);
+  ASSERT_EQ(res, 1);
+  ASSERT_EQ(out_len, 0);
+
+  // Get the tag
+  auto computed_tag = (uint8_t *) OPENSSL_malloc(tag.size());
+  res = EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, tag.size(),
+                            computed_tag);
+  ASSERT_NE(res, 0);
+
+  // Initialize the decrypt context
+  EVP_CIPHER_CTX *dctx = EVP_CIPHER_CTX_new();
+  ASSERT_TRUE(dctx);
+  res = EVP_DecryptInit_ex(dctx, cipher, NULL, key.data(), iv.data());
+  ASSERT_EQ(res, 1);
+
+  // Insert AAD
+  out_len = 0;
+  res = EVP_DecryptUpdate(dctx, NULL, &out_len, aad.data(), aad.size());
+  ASSERT_EQ(res, 1);
+  ASSERT_EQ(out_len, (int) aad.size());
+
+  // Insert ciphertext
+  auto computed_pt = (uint8_t *) OPENSSL_malloc(msg.size());
+  if (msg.size() > 0) {
+    out_len = 0;
+    res = EVP_DecryptUpdate(dctx, computed_pt, &out_len, computed_ct,
+                            ct.size());
+    ASSERT_EQ(res, 1);
+    ASSERT_EQ(out_len, (int) msg.size());
+  }
+
+  // Set the tag
+  res = EVP_CIPHER_CTX_ctrl(dctx, EVP_CTRL_AEAD_SET_TAG, tag.size(),
+                            tag.data());
+  ASSERT_NE(res, 0);
+
+  // Finish decryption
+  out_len = 0;
+  res = EVP_DecryptFinal(dctx, NULL, &out_len);
+
+  // Tag validation, a valid result should end with a valid tag
+  ASSERT_EQ(res, result.IsValid() ? 1 : 0);
+  ASSERT_EQ(out_len, 0);
+
+  // Valid results should match the KATs
+  if (result.IsValid()) {
```

Comment:
Did you look at `RunWycheproofTestCase` in aead_test.cc? That flow is easier to follow, there is no need to do a lot of this test if the test is supposed to fail.